### ISSUE TITLE
Possible bug in latent vector loss calculation?

### ISSUE THREAD

**Issue by walmsley** [lucidrains/big-sleep#34]
I'm confused by this, and wondering if it could be a bug? It seems as though `latents` is of size (32,128), which means that `for array in latents:` iterates 32 times. However, the results from these iterations aren't stored anywhere, so they are at best a waste of time and at worst causing a miscalculation. Perhaps the intention was to accumulate the kurtoses and skews for each array in latents, and then computing `lat_loss` using all the accumulated values?

```        
for array in latents:
    mean = torch.mean(array)
    diffs = array - mean
    var = torch.mean(torch.pow(diffs, 2.0))
    std = torch.pow(var, 0.5)
    zscores = diffs / std
    skews = torch.mean(torch.pow(zscores, 3.0))
    kurtoses = torch.mean(torch.pow(zscores, 4.0)) - 3.0

lat_loss = lat_loss + torch.abs(kurtoses) / num_latents + torch.abs(skews) / num_latents
```

Occurs at https://github.com/lucidrains/big-sleep/blob/main/big_sleep/big_sleep.py#L211

**Comment by walmsley** [comment]
Digging around further in the loss calculation code, I'm also curious about two other points:

1. Is this supposed to be `torch.abs(torch.mean(latents, dim=1)).mean()`? Otherwise I believe the outer mean is redundant because the inner mean has already reduced everything single value.
https://github.com/lucidrains/big-sleep/blob/3531538b167542a23081b99b0eba563173c990b5/big_sleep/big_sleep.py#L199

2.
The class loss term processes `soft_one_hot_classes` of size (32,1000) but if I'm reading it correctly, the zero index `[0]` extracts only the first of 32 arrays from the output of the `topk()` operation in `topk(soft_one_hot_classes, ..., dim=1)[0]`.
a) Is it intentional that the loss only cares about the first array of `soft_one_hot_classes` and ignores the other 31?
b) If so, then the topk operation looks like it could be made faster by only operating on `topk(soft_one_hot_classes[0], ...)` instead of `topk(soft_one_hot_classes, ..., dim=1)[0]`
https://github.com/lucidrains/big-sleep/blob/3531538b167542a23081b99b0eba563173c990b5/big_sleep/big_sleep.py#L212

**Comment by lucidrains** [comment]
@walmsley Hi Will! You caught a bug in regards to your first comment! I've fixed it in 0.5.1 :pray: 

For your second comment, I double checked the original colab from Ryan, and that's what he has there, so you may want to redirect your question to him (he originally devised this technique)

As for #2, topk actually returns a tuple, the first element being the actual topk values, and the second being the topk indices. The [0] is there to capture the values, not for referencing the first latent

**Comment by lucidrains** [comment]
@walmsley going to go with your suggestion for #1, because I think you are right :)

**Comment by walmsley** [comment]
That fully addresses the above points, thanks!

BUT also perhaps most significantly, there is a deeper possible bug that I'm curious about. I was trying to understand why `num_latents` is set to 32 here:
https://github.com/lucidrains/big-sleep/blob/31fa846e0e300515360fbeb67a0315f20505fd59/big_sleep/big_sleep.py#L89-L92

Diving deeper, it seems as though these 32 different vectors are only actually used within `cond_vector` here:
https://github.com/lucidrains/big-sleep/blob/31fa846e0e300515360fbeb67a0315f20505fd59/big_sleep/biggan.py#L510
and
https://github.com/lucidrains/big-sleep/blob/31fa846e0e300515360fbeb67a0315f20505fd59/big_sleep/biggan.py#L520

I debugged the loop surrounding line 520 above (using the current 512px BigGAN model), and found that the model only actually contains 15 layers; of those 15, only 14 of those layers are GenBlock layers which trigger line 520.

The result is that of the 32 latent vectors we create, only indices {0,1,2,3,4,5,6,7,8,10,11,12,13,14,15} are actually ever used. This wouldn't be a problem, except that the remaining 16 unused latent vectors may still be influencing the loss calculation. I'm still trying to work out whether their influence on the loss calculation is significant enough to merit fixing this, because the fix would be slightly nontrivial as it varies depending on the size of BigGAN model chosen.

**Comment by lucidrains** [comment]
@walmsley it's not a big deal, because there's only like less than 3 BigGAN models, and can just store a dictionary of the GAN size -> num latents somewhere

**Comment by walmsley** [comment]
Okay so I currently believe that a solution could look like this:

1) count the number of GenBlock layers by obtaining `len(self.model.biggan.config.layers)` (e.g. 14 layers), and add 1 to account for the 0th index being used for line 510. AKA `num_latents = len(self.model.biggan.config.layers) + 1` which is typically `15`.
2) modify the loop around line 520 to avoid incrementing the index when we arrive at a non-GenBlock layer: https://github.com/lucidrains/big-sleep/blob/31fa846e0e300515360fbeb67a0315f20505fd59/big_sleep/biggan.py#L518-L523 
i.e. change it to something like:
```
        next_available_latent_index = 1
        for layer in self.layers:
            if isinstance(layer, GenBlock):
                z = layer(z, cond_vector[next_available_latent_index].unsqueeze(0), truncation)
                next_available_latent_index += 1
            else:
                z = layer(z)
```

**Comment by htoyryla** [comment]
I wonder how much it helps to calculate the skews and kurtosis for the relevant latents only, when all 32 latents are anyway used for the same loss here https://github.com/lucidrains/big-sleep/blob/31fa846e0e300515360fbeb67a0315f20505fd59/big_sleep/big_sleep.py#L198-L200

Came to think about this while experimenting on image search from BigGAN using the same code as a basis. Perhaps we should instead simply dimension the Latents object with the correct amount to begin with.

Also we are still missing the proper way to accumulate the skews and kurtoses from each latent. Probably simply add them up, or perhaps add up their absolute values.

**Comment by walmsley** [comment]
> Perhaps we should instead simply dimension the Latents object with the correct amount to begin with.

Yes, that was the intention, thanks for clarifying – `num_latents = len(self.model.biggan.config.layers) + 1` should be used to fix `num_latents` from 32->15 at the source, not just for the sake of kurtoses/skews.

> Also we are still missing the proper way to accumulate the skews and kurtoses from each latent
Note that this was fixed with release 0.5.1 as mentioned by @lucidrains

**Comment by htoyryla** [comment]
> Yes, that was the intention, thanks for clarifying – `num_latents = len(self.model.biggan.config.layers) + 1` should be used to fix `num_latents` from 32->15 at the source, not just for the sake of kurtoses/skews.
> 

To be precise, that must be done when the Latents object is instantiated, either by this default https://github.com/lucidrains/big-sleep/blob/31fa846e0e300515360fbeb67a0315f20505fd59/big_sleep/big_sleep.py#L92 or here https://github.com/lucidrains/big-sleep/blob/31fa846e0e300515360fbeb67a0315f20505fd59/big_sleep/big_sleep.py#L129-L132

**Comment by htoyryla** [comment]
> > Also we are still missing the proper way to accumulate the skews and kurtoses from each latent
> 
> Note that this was fixed with release 0.5.1 as mentioned by @lucidrains

I don't think anything was done to this. As far as I know, @lucidrains stated the code comes directly from Ryan (so I sent him a link to this discussion).

What I have currently is (but can't see any major effect... it might be that the whole skewness etc factor is not so critical here?).

```
skews = 0
    kurtoses = 0
    for array in lats.normu:
            mean = torch.mean(array)
            diffs = array - mean
            var = torch.mean(torch.pow(diffs, 2.0))
            std = torch.pow(var, 0.5)
            zscores = diffs / std
            skews += torch.abs(torch.mean(torch.pow(zscores, 3.0)))
            kurtoses +=  torch.abs(torch.mean(torch.pow(zscores, 4.0)) - 3.0)
```

**Comment by htoyryla** [comment]
> > Yes, that was the intention, thanks for clarifying – `num_latents = len(self.model.biggan.config.layers) + 1` should be used to fix `num_latents` from 32->15 at the source, not just for the sake of kurtoses/skews.
> 
> To be precise, that must be done when the Latents object is instantiated, either by this default
> 
> https://github.com/lucidrains/big-sleep/blob/31fa846e0e300515360fbeb67a0315f20505fd59/big_sleep/big_sleep.py#L92
> 
> or here
> https://github.com/lucidrains/big-sleep/blob/31fa846e0e300515360fbeb67a0315f20505fd59/big_sleep/big_sleep.py#L129-L132

Or just use latents[:num_latents] for the mean and std, maybe.

**Comment by walmsley** [comment]
@htoyryla the skew/kurtosis fix was subtle, it's just the change in indentation of line 211 here: https://github.com/lucidrains/big-sleep/commit/226b97398b1a0913d0656120357514fff3bad342#diff-a32d425a1d65b549cda9588699a004a9d283f46d0623256309606cc74f8d3dd8R211

**Comment by htoyryla** [comment]
> @htoyryla the skew/kurtosis fix was subtle, it's just the change in indentation of line 211 here: [226b973#diff-a32d425a1d65b549cda9588699a004a9d283f46d0623256309606cc74f8d3dd8R211](https://github.com/lucidrains/big-sleep/commit/226b97398b1a0913d0656120357514fff3bad342#diff-a32d425a1d65b549cda9588699a004a9d283f46d0623256309606cc74f8d3dd8R211)

I see. I was only looking at the newest commit, with your name mentioned, did not realise this commit was also so recent.

Looks like it is equivalent to my solution.

**Comment by htoyryla** [comment]
BTW, my interest in this loss calculation comes from my ongoing experiment to skip the one-hot coded class vector in BigGAN here https://github.com/lucidrains/big-sleep/blob/31fa846e0e300515360fbeb67a0315f20505fd59/big_sleep/biggan.py#L574-L575 and instead use a 128 element class vector. I modified the loss to use similar code for the class vector as for the latents, and then I noticed the loss was not working correctly at all, but pushing the values to a wrong direction.

**Comment by walmsley** [comment]
Created new PR with proposed final fix @ https://github.com/lucidrains/big-sleep/pull/35

Overall status of 4 possible bugs mentioned in this issue:

- [X] kurtosis/skew accumulation (fixed in release 0.5.1)
- [X] latent loss mean(dim=1) (fixed in release 0.5.2)
- [X] class los topk[0] (not a bug, no fix needed)
- [ ] num_latents 32 -> 15 (PR filed pending review)
